examples/cpp/pyperf/PyPerfNativeStackTrace.cc

```diff
@@ -0,0 +1,152 @@
+  NativeStackTrace::stack = raw_stack;
+  NativeStackTrace::stack_len = stack_len;
+  NativeStackTrace::ip = ip;
+  NativeStackTrace::sp = sp;
+
+  unw_accessors_t my_accessors = _UPT_accessors;
+  my_accessors.access_mem = NativeStackTrace::access_mem;
+  my_accessors.access_reg = NativeStackTrace::access_reg;
+
+  unw_addr_space_t as = unw_create_addr_space(&my_accessors, 0);
+  void *upt = _UPT_create(pid);
+  if (!upt) {
+    logInfo(1, "Failed to _UPT_create\n");
+    this->symbols.push_back(std::string("[Error]"));
+    errors++;
+    return;
+  }
+
+  unw_cursor_t cursor;
+  int res = unw_init_remote(&cursor, as, upt);
+  if (res) {
+    logInfo(1, "Failed to unw_init_remote: %s\n", unw_strerror(res));
+    this->symbols.push_back(std::string("[Error]"));
+    errors++;
+    return;
+  }
+
+  do {
+    unw_word_t offset;
+    char sym[256];
+
+    if (unw_get_proc_name(&cursor, sym, sizeof(sym), &offset) == 0) {
+      this->symbols.push_back(std::string(sym));
+    } else {
+      unw_word_t ip;
+      unw_get_reg(&cursor, UNW_REG_IP, &ip);
+      logInfo(2,
+              "0x%lx -- error: unable to obtain symbol name for this frame "
+              "(SP=0x%lx)\n",
+              ip, NativeStackTrace::sp);
+      this->symbols.push_back(std::string("(missing)"));
+      errors++;
+      break;
+    }
+
+    if (memcmp(sym, "_PyEval_EvalFrameDefault",
+                sizeof("_PyEval_EvalFrameDefault")) == 0 ||
+        memcmp(sym, "PyEval_EvalFrameEx", sizeof("PyEval_EvalFrameEx")) == 0)
+        {
+      break;
+    }
+  } while (unw_step(&cursor) > 0);
+
+  _UPT_destroy(upt);
+  return;
+}
+
+int NativeStackTrace::access_reg(unw_addr_space_t as, unw_regnum_t regnum,
+                                 unw_word_t *valp, int write, void *arg) {
+  if (regnum == UNW_REG_SP) {
+    if (write) {
+      logInfo(2, "Libunwind attempts to write to SP\n");
+      return -UNW_EINVAL;
+    }
+
+    *valp = NativeStackTrace::sp;
+    return 0;
+  }
+  else if (regnum == UNW_REG_IP) {
+    if (write) {
+      logInfo(2, "Libunwind attempts to write to IP\n");
+      return -UNW_EINVAL;
+    }
+
+    *valp = NativeStackTrace::ip;
+    return 0;
+  }
+  else {
+    logInfo(2, "Libunwind attempts to %s regnum %d\n", write ? "write" : "read", regnum);
+    return -UNW_EBADREG;
+  }
+}
+
+int NativeStackTrace::access_mem(unw_addr_space_t as, unw_word_t addr,
+                                 unw_word_t *valp, int write, void *arg) {
+  if (write) {
+    logInfo(2, "Libunwind mem write attempt\n");
+    return -UNW_EINVAL;
+  }
+
+  const unw_word_t stack_start = NativeStackTrace::sp & ~(getpagesize() - 1);
+  const unw_word_t stack_end = stack_start + NativeStackTrace::stack_len;
+
+  if (addr >= NativeStackTrace::sp && addr < stack_end) {
+    memcpy(valp, &stack[addr - stack_start], sizeof(*valp));
+    return 0;
+  } else if ((addr >= stack_end && addr < stack_end + getpagesize() * 2) ||
```

Comment:
Do you have any smarter way to identify "stack" accesses & block them? This method will break if a function in the trace has a very big stack - we'll skip too far down the stack and use `process_vm_readv`.

If no better way, we can increase this "red zone" around the stack we do bring.

Reply:
Not that I know of. Maybe I can get the boundaries of the stack by finding the VMA corresponding to SP, assuming this VMA will never be shared with any other memory section that libunwind might try to access, but it seems like an overkill.

Reply:
Okay. Let's increase the redzone to, say, 64 pages (deny reads 64 pages from either direction)

Reply:
The whole stack will never be that big, doesn't it? I think 32 pages would suffice.

Reply:
32 pages? It can definitely span across more than 32 pages. The default virtual size is 8mb (see https://unix.stackexchange.com/a/280865/387108).

Anyway, 32 / 64 don't matter much. 32 is also okay.

Reply:
Can update to either

Reply:
Interesting. I wrote it because I saw 21 pages in some `/proc/maps` I looked at